Model/Url/GetUrlsForProducts.php

```diff
@@ -47,34 +41,40 @@ public function __construct(
      */
     public function execute(array $productIds, int $storeId): array
     {
-        $productCollection = $this->productCollectionFactory->create();
-        // do not ignore out of stock products
-        $productCollection->setFlag('has_stock_status_filter', true);
-        // if array of product ids is empty, just load all products
-        if (!empty($productIds)) {
-            $productCollection->addIdFilter($productIds);
-        }
-        $productCollection->setStoreId($storeId);
-        $productCollection->addUrlRewrite();
-
         try {
             /** @var Store $store */
             $store = $this->storeManager->getStore($storeId);
         } catch (NoSuchEntityException $e) {
             return [];
         }
 
+        $useProductCanonical = $this->prerenderConfigHelper->isUseProductCanonicalUrlEnabled($storeId);
+
+        $findByData = [
+            UrlRewrite::ENTITY_TYPE => Rewrite::ENTITY_TYPE_PRODUCT,
+            UrlRewrite::STORE_ID => $storeId,
+            UrlRewrite::ENTITY_ID =>  $productIds
+        ];
+
+        $urlRewrites = $this->urlFinder->findAllByData($findByData);
+
         $this->emulation->startEnvironmentEmulation($storeId);
         $urls = [];
-        /** @var Product $product */
-        foreach ($productCollection as $product) {
-            $urlPath = $product->getData('request_path');
-            if (empty($urlPath)) {
+
+        foreach ($urlRewrites as $urlRewrite) {
+            if (empty($urlRewrite->getRequestPath())) {
+                continue;
+            }
+
+            // Ignore the product URL with category path.
+            if ($urlRewrite->getMetadata() && $useProductCanonical) {
```

Comment:
Let's switch these around so that if config is disabled, it will never check for metadata e.g.
```
if ($useProductCanonical && $urlRewrite->getMetadata()) {
```